articles/ros2_threat_model.md

```diff
@@ -0,0 +1,3291 @@
+  </tr>
+
+  <tr><th colspan="29">Embedded / Software / Communication / Short-Range
+Communication (e.g. Bluetooth)</th></tr>
+
+  <tr>
+  <td>Attacker executes arbitraty code using a short-range communication
+protocol vulnerability.</td>
+  <td class="danger">✘</td>
+  <td class="success">✓</td>
+  <td class="success">✓</td>
+  <td class="success">✓</td>
+  <td class="success">✓</td>
+  <td class="success">✓</td>
+  <td class="danger">3</td>
+  <td class="warning">2</td>
+  <td class="success">1</td>
+  <td class="success">1</td>
+  <td class="danger">3</td>
+  <td>10</td>
+  <td class="success">✓</td>
+  <td class="success">✓</td>
+  <td class="success">✓</td>
+  <td class="danger">✘</td>
+  <td class="danger">✘</td>
+  <td class="danger">✘</td>
+  <td class="danger">✘</td>
+  <td class="danger">✘</td>
+  <td class="success">✓</td>
+  <td class="success">✓</td>
+  <td class="success">✓</td>
+  <td class="success">✓</td>
+  <td class="success">✓</td>
+  <td class="success">✓</td>
+  <td>
+    <ul>
+      <li>Communications protocols should be disabled if unused (by using e.g.
+rfkill).</li>
+      <li>Binaries and libraries required to support short-range communications
+should be kept up-to-date.</li>
+    </ul>
+  </td>
+  <td><a href="http://dl.acm.org/citation.cfm?id=2028067.2028073">Checkoway,
+Stephen, Damon McCoy, Brian Kantor, Danny Anderson, Hovav Shacham, Stefan
+Savage, Karl Koscher, Alexei Czeskis, Franziska Roesner, and Tadayoshi Kohno.
+“Comprehensive Experimental Analyses of Automotive Attack Surfaces.” In
+Proceedings of the 20th USENIX Conference on Security, 6–6. SEC’11.
+Berkeley, CA, USA: USENIX Association, 2011.</a></td>
+  </tr>
+
+  <tr><th colspan="29">Embedded / Software / OS &amp; Kernel</th></tr>
+
+  <tr>
+    <td>An attacker compromises the real-time clock to disrupt the kernel RT
+scheduling guarantees.</td>
+    <td class="danger">✘</td>
+    <td class="danger">✘</td>
+    <td class="danger">✘</td>
+    <td class="danger">✘</td>
+    <td class="success">✓</td>
+    <td class="danger">✘</td>
+    <td class="danger">3</td>
+    <td class="warning">2</td>
+    <td class="success">1</td>
+    <td class="danger">3</td>
+    <td class="warning">2</td>
+    <td>11</td>
+    <td class="success">✓</td>
+    <td class="danger">✘</td>
+    <td class="success">✓</td>
+    <td class="danger">✘</td>
+    <td class="danger">✘</td>
+    <td class="danger">✘</td>
+    <td class="danger">✘</td>
+    <td class="danger">✘</td>
+    <td class="success">✓</td>
+    <td class="success">✓</td>
+    <td class="success">✓</td>
+    <td class="danger">✘</td>
+    <td class="danger">✘</td>
+    <td class="danger">✘</td>
+    <td>
+      <ul>
+        <li>Hardened kernel (prevent dynamic loading of kernel modules)</li>
+        <li>Ensure only trustable kernels are used (e.g. Secure Boot)</li>
+        <li>/boot should not be accessible by robot processes</li>
+      </ul>
```

Comment:
Outside of the Kernel modules the robot is likely running an implementation of NTP to modify the RTC on the device.  A successful attack on NTP enables a number of other attacks. From enabling the use of expired keys, distortion of time data when used with sensors or other components in the system, or hampering coordination with remote components with correct time data.

Mitigations would include using NTP with encryption or using multiple NTP pools to detect modified times.

Reply:
Agree with that, having no hardware RTC on the Raspberry pi supposses a problem, as no valid time exists at boot. Depending on the implementation, it stores the latest known valid time or goes straight to 0 epoch. I would suggest the addition of a dedicated hardware RTC as a mitigation as well. Adding limitations to the maximum time difference to be taken as valid from the NTP server would also help.

Reply:
@cratliff @olaldiko I opened https://github.com/ros2/design/pull/229 as a follow-up PR to address this feedback. I will close the discussion here - your feedback on this PR would be greatly appreciated!